Filter log display by user

itutu-tienday opened the pull request:
## Content

In the Log Viewer, a function has been implemented to filter the log display so that only the user's own log is displayed.

### Remarks

Initially, I considered switching the output log file for each user, but due to issues with processing complexity and stability, this PR attempts a simpler solution.
- Each user's log is recorded in the same log file, but with unique information that identifies the user.
- The Log Viewer filters and displays the log by user, using user-specific information in the log as a key.

## Specification

### Architecture

```mermaid
graph LR
    subgraph FrontEnd
        Screens[each Screens]
        LogViewer[log viewer]
    end

    subgraph BackEnd
        APIs[each APIs]
        
        subgraph SnakemakeRunModules["Snakemake run modules"]
            DataNode["data node <br> (data.py)"]
            AlgoNode["algo node <br> (func.py)"]
        end
        
        ContextVars[("contextvars <br> (python module)")]
        Logger[logger]
        LogFile[(log file)]
        LogAPI["API (GET logs)"]
    end

    %% FrontEnd to BackEnd connections
    Screens -->|request| APIs

    %% BackEnd internal connections
    APIs -->|set client_id| ContextVars
    ContextVars -->|get client_id| Logger
    ContextVars -->|get client_id| SnakemakeRunModules
    ContextVars -->|get client_id| LogAPI
    APIs -->|logging| Logger
    SnakemakeRunModules -->|logging| Logger
    
    Logger -->|"logging <br> (with client_id)"| LogFile
    LogFile -->|"get logs <br> (filter by client_id)"| LogAPI
    
    %% BackEnd to FrontEnd connections
    LogViewer -->|request logs| LogAPI
    LogAPI -->|response logs| LogViewer
```

### Specification Explanation

- The client_id uses a hash value based on the Firebase UID, etc., allowing for user identification while recording anonymous information in the log.
  - In cases where the client_id cannot be obtained (such as before signing in or in standalone mode), `client:default` is recorded.

- It has been confirmed that the client_id cannot be obtained in some log output locations.
  - Example
    - Logging of conda activate process in snakemake module
      *The logging for "Activating conda environment" is set to `client:default`
        ```
        2025-10-14 16:50:02,580 INFO     [snakemake.logging] (pid:70659) (client:4e782f44eac52685) text_handler():510 -
        2025-10-14 16:50:02,583 WARNING  [snakemake.logging] (pid:70659) (client:default) text_handler():578 - Activating conda environment: ../../../../barebone-studio/.snakemake/conda/31702a884cc2afe3a5cece2e88881218_
        ``` 
    - Reason:
      - Snakemake uses a thread pool internally, and there appear to be some paths where contextvars cannot be shared between different threads.

## References

- arayabrain/optinist-for-cloud#99

## Testcase

### multiuser mode

#### 単一ユーザーでの確認

- [ ] log file の内容確認（client_id の記録）
  - [ ] sign in 前の画面（sing in form）へのアクセス
    - [ ] log file (logs/studio.log) に、以下のようなLogが記録されていること
      > 2025-10-14 16:31:19,195 [32mINFO[0m:     [uvicorn.access] (pid:69668) (client:default) send():478 - 127.0.0.1:63986 - "GET /is_standalone HTTP/1.1" 200
        - `(client:{client_id})` の {client_id} に、固定文字列 "default" が記録されている
  - [ ] sign in 後の画面へのアクセス
    - [ ] log file (logs/studio.log) に、以下のようなLogが記録されていること
      > 2025-10-14 15:20:16,186 [32mINFO[0m:     [uvicorn.access] (pid:64405) (client:4e782f44eac5) send():478 - 127.0.0.1:62774 - "GET /experiments/1 HTTP/1.1" 200
        - `(client:{client_id})` の {client_id} に、16文字のハッシュ値が記録されている

- [ ] Log Viewer の動作確認  
  ※Worfklow画面で、Log Viewer を表示する ("Logs" icon を Click) 
  <img width="372" height="75" alt="image" src="https://github.com/user-attachments/assets/bc3908c4-a267-4eec-b6f0-4ac429e08c29" />
  - [ ] Log Viewerには、自身の操作で記録されたLogのみが表示されること
    - `(client:{client_id})` に複数の {client_id} が混在して表示されていないこと
    - 生log fileと比較確認
  - [ ] その他、Log Viewer を操作しても、自身の操作以外のLogが表示されていないこと
    - [ ] Workflowの実行 … リアルタイムで自身のWorkflowのLogのみが表示されること
    - [ ] Log Level Filter  … Filterが機能しつつ、自身の操作のLogのみが表示されること
    - [ ] Log histroy scroll back  … 自身の操作のLogのみが表示されること

#### 複数ユーザーでの確認

※別のユーザーで2名同時にログインし、Log Viewer の表示を確認する

- [ ] log file の内容確認
  - [ ] 2名のユーザー間で、log file に異なる {client_id} が記録されていること
- [ ] Log Viewer の動作確認
  - [ ] 2名のユーザー間で、Log Viewerには、自身の操作で記録されたLogのみが表示されること
    - 生log fileと比較確認

### standalone mode

- [ ] log file の内容確認
  - [ ] log file (logs/studio.log) に、以下のようなLogが記録されていること
    > 2025-10-14 16:39:27,833 [32mINFO[0m:     [uvicorn.access] (pid:69668) (client:default) send():478 - 127.0.0.1:64045 - "GET /workspaces?offset=0&limit=50 HTTP/1.1" 401
      - `(client:{client_id})` の {client_id} に、固定文字列 "default" が記録されている

- [ ] Log Viewer の動作確認  
      <img width="372" height="75" alt="image" src="https://github.com/user-attachments/assets/bc3908c4-a267-4eec-b6f0-4ac429e08c29" />
  - [ ] Log Viewerには、すべてのLogが表示されること
    - `(client:{client_id})` の {client_id} によらず、すべてのLogが表示されていること
    - 生log fileと比較確認
  - [ ] その他、Log Viewer を操作しても、すべてのLogが表示されていること
    - [ ] Workflowの実行 … リアルタイムで すべてのLogが表示されること
    - [ ] Log Level Filter  … Filterが機能しつつ、すべてのLogが表示されること
    - [ ] Log histroy scroll back  … すべてののLogが表示されること

Commit:
Filter logging by user) first commit

Commit:
Filter logging by user) Supports user_id filtering in log viewer

Commit:
Filter logging by user) Supports user_id logging for snakemake process

milesAraya commented:
> 2025-10-15 09:50:31,364 [32mINFO[0m:     [uvicorn.error] (8566) startup():60 - Application startup complete.
> 2025-10-15 09:50:31,367 [32mINFO[0m:     [uvicorn.error] (8566) _log_started_message():218 - Uvicorn running on [1mhttp://127.0.0.1:8001[0m (Press CTRL+C to quit)
> 2025-10-15 09:54:38,320 [32mINFO[0m:     [uvicorn.access] (8566) send():478 - 127.0.0.1:50713 - "POST /auth/login HTTP/1.1" 200

- What does "[32mINFO[0m: " mean? Can this be removed?
- "uvicorn.error"?

Console log becomes a bit busy as log checked often

> 2025-10-15 10:01:42,153 INFO:     [uvicorn.access] (8566) send():478 - 127.0.0.1:50954 - "GET /logs?offset=25331&search=&reverse=false HTTP/1.1" 200
> 2025-10-15 10:01:45,155 INFO:     [uvicorn.access] (8566) send():478 - 127.0.0.1:50956 - "GET /logs?offset=25489&search=&reverse=false HTTP/1.1" 200
> 2025-10-15 10:01:48,152 INFO:     [uvicorn.access] (8566) send():478 - 127.0.0.1:50956 - "GET /logs?offset=25647&search=&reverse=false HTTP/1.1" 200
> 2025-10-15 10:01:51,150 INFO:     [uvicorn.access] (8566) send():478 - 127.0.0.1:50958 - "GET /logs?offset=25805&search=&reverse=false HTTP/1.1" 200
> 2025-10-15 10:01:54,148 INFO:     [uvicorn.access] (8566) send():478 - 127.0.0.1:50958 - "GET /logs?offset=25963&search=&reverse=false HTTP/1.1" 200
> 2025-10-15 10:01:57,150 INFO:     [uvicorn.access] (8566) send():478 - 127.0.0.1:50961 - "GET /logs?offset=26121&search=&reverse=false HTTP/1.1" 200
> 2025-10-15 10:02:00,149 INFO:     [uvicorn.access] (8566) send():478 - 127.0.0.1:50965 - "GET /logs?offset=26279&search=&reverse=false HTTP/1.1" 200
> 2025-10-15 10:02:03,142 INFO:     [uvicorn.access] (8566) send():478 - 127.0.0.1:50965 - "GET /logs?offset=26437&search=&reverse=false HTTP/1.1" 200

itutu-tienday commented:
@milesAraya 
> What does "�[32mINFO�[0m: " mean? Can this be removed?

Characters like `[32m` are terminal escape sequences, controlled by python logging. (This has been the default behavior since before this PR.)
- See: https://en.wikipedia.org/wiki/ANSI_escape_code

> "uvicorn.error"?

This is also unrelated to this PR and is a log that has been output since the beginning.
(It would be good to investigate further though.)

> Console log becomes a bit busy as log checked often

The above log (GET /logs) is filtered by default so that it does not appear in the log viewer.
(I think there is no problem with logging it to the log file.)

Commit:
Filter logging by user) Replaced user_id with client_id

milesAraya commented on studio/app/common/core/snakemake/snakemake_executor.py:
```diff
@@ -24,11 +24,14 @@
 
 def snakemake_execute(workspace_id: str, unique_id: str, params: SmkParam):
+    # Get client_id from current context to pass to subprocess
+    client_id = AppLogger.get_client_id()
```
Previously the logger was set once at the start of the script 
> logger = AppLogger.get_logger()

Now we need to add for every function? 
Could client_id = AppLogger.get_client_id() not be included somehow in AppLogger.get_logger()? 
If this client_id = AppLogger.get_client_id() is required for every function, could you add a section to Developer documentation explaining the recommended useage?
Maybe [here](https://optinist.readthedocs.io/en/latest/for_developers/index.html)

itutu-tienday replied:
@milesAraya 
First of all, as explained in the code comments, the execution of `client_id = AppLogger.get_client_id()` is limited to a few places.

1. The place where snakemake yaml is generated
    - studio/app/common/core/workflow/workflow_runner.py
2. The place where processing spans a different thread (the place where the snakemake process is executed with a ProcessPoolExecutor)
    - studio/app/common/core/snakemake/snakemake_executor.py
    - studio/app/optinist/core/edit_ROI/edit_ROI.py

Currently, there are only three places where the above occurs.

Since these are areas that are not often customized by users, if we were to add more documentation, it would be more appropriate to expand the comments in the source code rather than using readthedocs.

What do you think? Please let me know if you have any comments.

itutu-tienday replied:
Note) The current code seems to be working fine, but I will consider ways to simplify it a bit more.

itutu-tienday replied:
@milesAraya 
The following commit applies code refactoring:
- ef4e2e2

#### Updates
- Added logger_context_helpers.py
  - In snakemake execution related modules, I use the functions in this helper (which wrap AppLogger methods) to clarify the processing names.
  - In addition, by providing a decorator (with_client_id_context), AppLogger operations (such as AppLogger.set_client_id) are hidden and automated.

Please check it out.

Commit:
Filter logging by user) Refactoring to simplify the code
- Added logger_context_helpers.py
  - In snakemake execution related modules, we use the functions in this helper (which wrap AppLogger methods) to clarify the processing names.
  - In addition, by providing a decorator (with_client_id_context), AppLogger operations (such as AppLogger.set_client_id) are hidden and automated.

milesAraya reviewed (approved):
Looks good!

tsuchiyama-araya reviewed (approved):
LGTM　動作確認もOKです
ありがとうございます。
